core/src/main/java/com/linecorp/armeria/common/HttpMessageAggregator.java

```diff
@@ -0,0 +1,284 @@
+                        } else {
+                            foundNonInformationalHeaders = true;
+                            responseHeaders = headers0;
+                            dataStart = i + 1;
+                            break;
+                        }
+                    } else if (httpObject instanceof HttpHeaders) {
+                        // Append to the last informational headers
+                        final HttpHeaders headers0 = (HttpHeaders) httpObject;
+                        final int lastIdx = informationals.size() - 1;
+                        informationals.set(lastIdx,
+                                           informationals.get(lastIdx).withMutations(h -> h.add(headers0)));
+                    }
+                }
+            }
+
+            if (!foundNonInformationalHeaders) {
+                for (int i = dataStart; i < size; i++) {
+                    PooledObjects.close(objects.get(i));
+                }
+                throw new IllegalStateException(
+                        "An aggregated message does not have a non-informational headers.");
+            }
+
+            int dataEnd = -1;
+            int contentLength = 0;
+            HttpHeaders trailers = HttpHeaders.of();
+            for (int i = dataStart; i < size; i++) {
+                final HttpObject httpObject = objects.get(i);
+                if (dataEnd > 0) {
+                    // Ignore 'HttpObject's written after a trailers.
+                    PooledObjects.close(httpObject);
+                } else {
+                    if (httpObject instanceof HttpHeaders) {
+                        trailers = (HttpHeaders) httpObject;
+                        dataEnd = i;
+                    } else {
+                        contentLength += ((HttpData) httpObject).length();
+                    }
+                }
+            }
+            if (dataEnd == -1) {
+                dataEnd = size;
+            }
+
+            if (contentLength == 0) {
+                return AggregatedHttpResponse.of(informationals, responseHeaders, HttpData.empty(), trailers);
+            }
+
+            final HttpData content = aggregateData(objects, contentLength, dataStart, dataEnd, alloc);
+            return AggregatedHttpResponse.of(informationals, responseHeaders, content, trailers);
+        }));
+    }
+
+    private static HttpData aggregateData(List<HttpObject> objects, int contentLength, int start, int end,
+                                          @Nullable ByteBufAllocator alloc) {
+        if (alloc != null) {
+            final ByteBuf merged = alloc.buffer(contentLength);
+            for (int i = start; i < end; i++) {
+                try (HttpData data = (HttpData) objects.get(i)) {
+                    final ByteBuf buf = data.byteBuf();
+                    if (!data.isEmpty()) {
+                        merged.writeBytes(buf, buf.readerIndex(), data.length());
+                    }
```

Comment:
Where do we release `data`?

Reply:
Actually, I copied the original code. The `data` is closed by the enclosing try-with-resource block. 😆
```
try (HttpData data = (HttpData) objects.get(i)) {
```

If it is not readable, I can update it.